src/runtime-rs/crates/resource/src/network/utils/link/create.rs

```diff
@@ -0,0 +1,150 @@
+    pub ifr_data: *mut libc::c_char,
+}
+
+#[repr(C)]
+struct CreateLinkReq {
+    pub ifr_name: IfName,
+    pub ifr_ifru: CreateLinkIfru,
+}
+
+impl CreateLinkReq {
+    pub fn from_name(name: &str) -> io::Result<CreateLinkReq> {
+        let mut req: CreateLinkReq = unsafe { mem::zeroed() };
+        req.set_name(name)?;
+        Ok(req)
+    }
+
+    pub fn set_name(&mut self, name: &str) -> io::Result<()> {
+        set_name!(self.ifr_name, name)
+    }
+
+    pub fn get_name(&self) -> io::Result<String> {
+        get_name!(self.ifr_name)
+    }
+
+    pub unsafe fn set_raw_flags(&mut self, raw_flags: libc::c_short) {
+        self.ifr_ifru.ifr_flags = raw_flags;
+    }
+}
+
+const DEVICE_PATH: &str = "/dev/net/tun";
+
+bitflags! {
+    pub struct TunTapFlags: libc::c_short {
+        const IFF_TUN = 0x0001;
+        const IFF_TAP = 0x0002;
+        const IFF_NAPI = 0x0010;
+        const IFF_NAPI_FRAGS = 0x0020;
+        const IFF_NO_PI = 0x1000;
+        /// this flag has no real effect
+        const IFF_ONE_QUEUE = 0x2000;
+        const IFF_VNET_HDR = 0x4000;
+        #[allow(overflowing_literals)]
+        const IFF_TUN_EXCL = 0x8000;
+        const IFF_MULTI_QUEUE = 0x0100;
+        const IFF_ATTACH_QUEUE = 0x0200;
+        const IFF_DETACH_QUEUE = 0x0400;
+        /// read-only flag
+        const IFF_PERSIST = 0x0800;
+        const IFF_NOFILTER = 0x1000;
+    }
+}
+
+ioctl_write_ptr!(tun_set_iff, b'T', 202, libc::c_int);
+ioctl_write_ptr!(tun_set_persist, b'T', 203, libc::c_int);
+
+#[derive(Clone, Copy, Debug)]
+pub enum LinkType {
+    #[allow(dead_code)]
+    Tun,
+    Tap,
+}
+
+pub fn create_link(name: &str, link_type: LinkType, queues: usize) -> Result<()> {
+    //let mut flags = TunTapFlags::IFF_NO_PI;
+    let mut flags = TunTapFlags::IFF_VNET_HDR;
+    flags.insert(match link_type {
+        LinkType::Tun => TunTapFlags::IFF_TUN,
+        LinkType::Tap => TunTapFlags::IFF_TAP,
+    });
+
+    let queues = if queues == 0 { 1 } else { queues };
+    if queues > 1 {
+        flags.insert(TunTapFlags::IFF_MULTI_QUEUE | TunTapFlags::IFF_NO_PI);
+    } else {
+        flags.insert(TunTapFlags::IFF_ONE_QUEUE);
+    };
+
+    // create queue 1
+    let mut files = vec![];
+    let (file, result_name) = create_queue(name, flags)?;
+    unsafe {
+        tun_set_persist(file.as_raw_fd(), &1)?;
+    }
+    files.push(file);
+
+    // create queue > 1
+    if queues > 1 {
+        for _ in 0..queues - 1 {
+            files.push(create_queue(&result_name, flags)?.0);
+        }
+    }
+
+    info!(sl!(), "create link with fds {:?}", files);
+    Ok(())
+}
+
+fn create_queue(name: &str, flags: TunTapFlags) -> Result<(File, String)> {
+    let path = Path::new(DEVICE_PATH);
+    let file = OpenOptions::new().read(true).write(true).open(&path)?;
+    let mut req = CreateLinkReq::from_name(name)?;
+    unsafe { req.set_raw_flags(flags.bits()) };
+    unsafe { tun_set_iff(file.as_raw_fd(), &mut req as *mut _ as *mut _) }?;
```

Comment:
Can we merge these two unsafe blocks?

Reply:
Thanks for your review has been updated.